File: 129. Sum Root to Leaf Numbers.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.res = 0
        self.preorder(0, root)
        return self.res
        
    def preorder(self, s, root):
        if not root:
            return
        if not root.left and not root.right:
            self.res += 10*s + root.val
            return
        self.preorder(10*s + root.val, root.left)
        self.preorder(10*s + root.val, root.right)

File: 129. Sum Root to Leaf Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int res = 0;
        preorder(res, 0, root);
        return res;
    }
    void preorder(int& res, int s, TreeNode* root)
    {
        if (root == NULL)
            return;
        if (root->left == NULL && root->right == NULL)
        {
            res += 10*s + root->val;
            return;
        }
        preorder(res, 10*s + root->val, root->left);
        preorder(res, 10*s + root->val, root->right);
    }
};
